README.md
# Controllino-CMP
This is a Arduino library (cpp and header file) written to provide a message interface on Controllino Mega devices.

There is a LabVIEW client available [LV-Controllino-CMP (https://github.com/CS-GSI/LV-Controllino-CMP.git)](https://github.com/CS-GSI/LV-Controllino-CMP.git).

# CMP - ControllinoMessageProtocol

This protocol is designed to communicate with controllino devices in a command based request-response pattern. It allows you to read/write controllino IOs (Digital, Analog, Relays) and volatile memories. CMP is a simple lightweight ASCII based protocol, all characters must be ASCII.

Implemented and partly tested for Controllino Mega. Differences to other Contrllino Devices have not been taken into account. 

___
```
CMP - ControllinoMessageProtocol (V01) 	Version 'A'

	'A': V01 - 'A' like First, 'A' like ASCII, 'A' like A-Z should be enough characters for future development
	You have to replace 'V' with 'A' in any exampla below to make it work!

Determining CMP version of your Controllino

	You may either send an invalid command and hope for Error VEE01 or simply send Set/Read Error.
	Sending invalid commands is possibly unsafe since you may not know how an unknown CMP version may react to it. 

CMP Short Description
	Command		|	Data D			|	Description						| 	Response
	-------------------------------------------------------------------------------------------------
	VSCNNDM		|	'O','P','I'		|	Set Config						|	Error or Command (echo)
	VSDNNDM		|	'H', 'L'		|	Set Digital						|	Error or Command (echo)
	VSENNDM		|	-				|	Set Error						|	Error or Command (echo)
	VSRNNDM		|	'H', 'L'		|	Set Relay						|	Error or Command (echo)
	VSVNNDM		|	Int16 (ASCII)*	|	Set Volatile Memory Register	|	Error or Command (echo)
	VRANN		|	-				|	Read Analog Input				|	
	VRCNN		|	-				|	Read Digitals IO Configuration	|	
	VRDNN		|	-				|	Read Digitals					|	
	VRENN		|	-				|	Read Error						|	
	VRINN		|	-				|	Read Digital Inputs				|	
	VRRNN		|	-				|	Read Relays						|	Error or ARRNND(DDDD)
	VRVNN		|	-				|	Read Volatile Memory Register 	|	

* Data is Mask!

CMP Message Frame: [ Version | Header | Data ]
	Version:		1 byte	[ V ]
	Header:			4 byte	[ R P NN ]	
	Data: 			N byte	[ D(D..) ]

	byte 0		V 	(Version)
					'A': Version A
	byte 1		R	(Request)
					'R': Read
					'S': Set
					'E': On Error Response
	byte 2		P	(Property)
					'A': Analog
					'C': Config
					'D': Digital
					'E': Error
					'I': Input
					'R': Relay
					'V': Volatile Memory Register
	byte 3,4 	N 	(Number)
					'00' to '99'
	byte 5		D 	(Data)
					ASCII Character
	byte 6...	M	(MASK)
					Additional Data such as Bitmasks

Request - Response Pattern:
	Controllino sends a response to any datagram received. 
	On error Controllino sends an error response.
	On SET-Requests controllino answers with a echo of the request.
		Minor differences: If Data has been set to default value the response contains the set value
	On READ-Requests Controllino answers with the command and data combined
	Responses also loop back any additional charcters that have been added to the Request
		This is simply due to how responses are generated (initially by making a copy of the request)

Note on strings:
	The current implementation can handle zero and non-zero terminated strings. 

Error are indicated by number. Known Errors:
	00: 	No Error
	01: 	CMP Version mismatch
	02: 	Digitals number out of scope
	03:		Relay number out of scope
	04:		Volatile memory access out of scope
	05:		Unknown Property
	06:		Analog number out of scope
	07:		Digital input number out of scope
	08:		Request unknown

Error Examples
	'x' equals to any invalid character
	'y' equals to any invalid or valid character
	Assuming Sizeof Volatile Memory is 50 and Contrllino Mega Configuration
	Send Command		|		Response
	-------------------------------------------------------
	x(y...)				| 		VEE01
	Vx(y...)			|		VEE08
	VSx(y...)			|		VEE05
	VRx(y...)			|		VEE05
	VSD86H(y...)		| 		VEE02
	VSC86P(y...)		|		VEE02
	VSR87H(y...)		|		VEE03
	VSV98123			|		VEE04
	VRV98(y...)			|		VEE04
	VRA98(y...)			|		VEE06
	VRI98(y...)			|		VEE07

Valid Protocols:
VSCNNDM(MMM)	Set Config (Number) IOMODE
			NN : 	'00' to '19'				(CONTROLLINO_D0 to CONTROLLINO_D19)
					'99' Special Character		
					Applys D to all IO Configs masked by bitmask M(MMM)
			D  : 	'O' (Output)	'P' (Input_Pullup) 	'I' (Input)*
					* Every Character except 'O' or 'P' equals to Input
			Response Message: 	(echo request) on success
								VEENN	on protocol error
			Example: 		Sender to controllino:	VSC02O	(Config Digital 02 as Output)
							controllino to Sender:	VSC02O	(Config Digital 02 as Output)
			Example: 		Sender to controllino:	VSC01P	(Config Digital 01 as Input with Pullup)
							controllino to Sender:	VSC01P	(Config Digital 01 as Input with Pullup)
			Example: 		Sender to controllino:	ASC01X	(Config Digital 19 as Default: Input)
							controllino to Sender:	ASC01I	(Config Digital 19 as Input)
			Example:		Sender to controllino:	ASC99P3	(Config Digital 00 and 01 as Input with Pullup)
							controllino to Sender:	ASC99P3	(Config Digital 00 and 01 as Input with Pullup)
			Example:		Sender to controllino:	ASC99X3	(Config Digital 00 and 01 as Default: Input)
							controllino to Sender:	ASC99X3	(Config Digital 00 and 01 Input with Pullup)

VSDNNDM(MMM)	Set Digital (Number) High/Low
			NN : 	'00' to '19'		(CONTROLLINO_D0 to CONTROLLINO_D19)
					'99' Special Character
					 Applys D to all Digitals masked by bitmask M(MMM)
			D  : 	'L' (Low)	'H' (High)*
					* Every Character except 'H' equals to Low
					Setting a digital output always implies a  'SCNNO' (setting as output)
			Response Message: 	(echo request) on success
								VEENN	on protocol error
			Example: 		Sender to controllino:	ASD00H	(Set Digital 00 to High)
							controllino to Sender:	ASD00H	(Set Digital 00 to High)
			Example:		Sender to controllino:	ASD00D	(Set Digital 00 to Default: Low)
							controllino to Sender:	ASD00L	(Set Digital 00 to Low)
			Example:		Sender to controllino:	ASD99H3	(Set Digital 00 and 01 to High)
							controllino to Sender:	ASD99H3	(Set Digital 00 and 01 to High)
			Also compare to examples above

VSENND			Set Error
			Set Error has no purpose
			You can use it to safely determine CMP version of Controllino or test an echo response
			Example:		Sender to controllino:	ASE
							controllino to Sender:	ASE
VSRNNDM(MMM)	Set Relay (Number) High/Low
			NN : 	'00' to '15'		(CONTROLLINO_R0 to CONTROLLINO_R15)
				 	'99' Special Character
				 	Applys D to all Relays masked by bitmask M(MMM)
			D  : 	'0' (Low)	'1' (High)*
					* Every Character except '1' equals to Low
			Response Message: 	VSRNND	(echo request) on success
								VEENN	on protocol error
			Example: 		Sender to controllino:	VSR02H	(Set Relay 02 to High)
							controllino to Sender:	VSR02H	(Set Relay 02 to High)
			Example:		Sender to controllino:	VSR00L	(Set Relay 00 to Low)
							controllino to Sender:	VSR00L	(Set Relay 00 to Low)
			Also compare to examples above

VSVNNDQ(QQQ)	Set Volatile Memory Register (Number) Value
			NN : 	'00' to '50'		(Array volmem[SIZEOFVOLMEM])
			D  : 	sign '+' or '-'
			Q(QQQ): 16bit Integer value as ASCII String
			Note: No range check of Q (no under/overflowdetection of int16_t)
			Response Message: 	VSVNND	(echo request) on success
								VEENN	on protocol error
			Example: 		Sender to controllino:	VSV00z20		(volmem[0] = 20)
							controllino to Sender:	VSV00z20		(volmem[0] = 20)
			Example: 		Sender to controllino:	VSV03z-24O		(volmem[3] = -240)
							controllino to Sender:	VSV03z-240		(volmem[3] = -240)
			Also compare to examples above

VRANN			Read Analog Input (Number)
			NN : 	'00' to '15'		(CONTROLLINO_A0 to CONTROLLINO_A15)
					'99' Special Character
				 	Reads all analog inputs in sequence
			Response Message: 	VRANND to VRANNDDDD	on success
								D(DDD) : Analog Value as numeric ASCII characters corresponding to 16 bit unsigned int
								Only in case of NN == '99'
									RANDD(DD).D(DD)....D(DD) (16 Values separated by dot)
								VEENN	on protocol error
			Example: 		Sender to controllino:	VRA00		(Read Analog Input 00)
							controllino to Sender:	VRA00824	(Read Analog Input 00, Example Data: 824)
			Example:		Sender to controllino:	VRA99		(Read all Analog Inputs)
							controllino to Sender:	ARA990.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0	(Read all Analog Inputs, Example Data: all zero)
			Also compare to examples above

VRCNN			Read Digitals IO Configuration	(CONTROLLINO_D0 to CONTROLLINO_D19)
			NN : 	'00' to '19'		(CONTROLLINO_D0 to CONTROLLINO_D19)
					'99' Special Character
			Response Message: 	VRCNNM
									M:	ASCII Characters corresponding to IO Config ('O' , 'P' , 'I')
								Only in case of NN == '99'
									VRCMMMMMMMMMMMMMMMMMMMM on success (19 times M)
								VEENN	on protocol error
			Example: 		Sender to controllino:	VRC00	(Read IO Configuration of D0)
							controllino to Sender:	VRCOI 	(Read Input, Data: configured as Input)	
			Example:		Sender to controllino:	VRC99	(Read all IO Configuration)
							controllino to Sender:	VRCOOOOOOOOOOOOOOOOOOOO	(Read Input, Data: All Inputs configured as Output)	
			Also compare to examples above

VRDNN			Read Digitals	(CONTROLLINO_D0 to CONTROLLINO_D19)
			NN : 	'00' to '19'		(CONTROLLINO_D0 to CONTROLLINO_D19)
					'99' Special Character
			Response Message: 	VRDNNM
									M:	ASCII Characters corresponding to IO State ('H', 'L')
								Only in case of NN == '99'
									VRDD.D to VRDDDDDD.DDDDD	on success
									D(DDDD).D(DDDD) : 	Bitpattern Value as ASCII String corresponding to 32 bit unsigned int (two 16 bit unsigned int)
													Highword.Lowword
										Bit Meaning: 1 = Input High ; 0 = Input Low
										MSB1,...,LSB1,MSB0,...,LSB0 
										LSB0 : 	CONTROLLINO_D0
												...
										MSB0	CONTROLLINO_D15
										LSB1	CONTROLLINO_D16
												CONTROLLINO_D17
												CONTROLLINO_D18
												CONTROLLINO_D19
												0
												...
										MSB1   	0
								VEENN	on protocol error
			Example: 		Sender to controllino:	VRD000	(Read Digital 00)
							controllino to Sender:	VRD00H	(Read Digital 00, Data: High)
			Example:		Sender to controllino:	VRD99	(Read all Digitals)
							controllino to Sender:	VRD15.65535	(Read Input, Data: All Inputs High)	
			Also compare to examples above

VRENN			Read Error
			Read Error has no purpose
			You can use it to safely determine CMP version of Controllino or test an echo response
			Example:		Sender to controllino:	ARE
							controllino to Sender:	ARE

VRINN			Read Digital Inputs	(CONTROLLINO_I16, CONTROLLINO_I17, CONTROLLINO_I18, CONTROLLINO_IN0, CONTROLLINO_IN1)
			NN : 	16, 17, 18, 0, 1
					'99' Special Character
			Response Message: 	VRINNM
									M:	ASCII Characters corresponding to IO State ('H', 'L')
								Only in case of NN == '99'
									VRINND to VRINNDDD	on success
									D(DD) : Bitpattern as ASCII String corresponding to 8 bit unsigned int
										Bit Meaning: 1 = Input High ; 0 = Input Low
										LSB : 	CONTROLLINO_IN1
												CONTROLLINO_IN0
												CONTROLLINO_I18
												CONTROLLINO_I17
												CONTROLLINO_I16
												0
										MSB		0
								VEENN	on protocol error
			Example: 		Sender to controllino:	VRI16	(Read Input 16)
							controllino to Sender:	VRI16L	(Read Input, Data: All Inputs Low)	
			Example: 		Sender to controllino:	VRI99	(Read all Inputs)
							controllino to Sender:	VRI991 	(Read Input, Data: CONTROLLINO_IN1 High, rest is Low)

VRRNN			Read Relays	(CONTROLLINO_R0 to CONTROLLINO_R15)
			NN : 	'00' to '15'		(CONTROLLINO_R0 to CONTROLLINO_R15)
					'99' Special Character
			Response Message: 	VRRNNM
									M:	ASCII Characters corresponding to IO State ('H', 'L')
								Only in case of NN == '99'
									RRD to RRDDDDD	on success
									D(DDDD) : Bitpattern as ASCII String corresponding to 16 bit unsigned int
										Bit Meaning: 1 = Input High ; 0 = Input Low
										LSB : 	CONTROLLINO_R0
												...
										MSB		CONTROLLINO_R15
								VEENN	on protocol error
			Example: 		Sender to controllino:	VRR15	(Read Relay 15)
							controllino to Sender:	VRR15L	(Read Relays 15, Data: Relays Low)	
			Example: 		Sender to controllino:	VRR99	(Read Relays)
							controllino to Sender:	VRR990	(Read Input, Data: All relays low)	

VRVNN			Read Volatile Memory Register (Number)
			NN : '00' to '50'		(Array volmem[SIZEOFVOLMEM])
			Response Message: 	VRVNNDD to VRVNNDDDDDD on success
								DD(DDDD) : 16 bit Memory register value as ASCII character string
										: first character is the sign ('+' or '-')
								VEENN	on protocol error
			Example: 			Sender to controllino:	VRV01		(Read memory element 01)					
								controllino to Sender:	VRV01+10	(Read memory element 01, Data: 10)
			Example: 			Sender to controllino:	VRV05		(Read memory element 05)					
								controllino to Sender:	VRV06-32767	(Read memory element 01, Data: -32767)
```


Controllino_Firmware_UDP_Remote_IO/CHANGELOG.txt
#################### Controllino CMP Library ####################
## [0.1.0] - 2017-12-20
### Added
- Inital Release


#########################################################################
#########################################################################
#########################################################################

# Controllino CMP Library
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Time Format: YYYY-MM-DD

Major.Minor.Patch
Major 
	Erhebliche Änderungen im Funktionsumfang und umfangreiche Neuerungen auf Basis der Vorgänger. 
	Eine Vorgängerversion sollte mindestens einmalig erfolgreich z.B. während einer Strahlzeit eingesetzt worden sein bevor sie als Grundlage für die nächste Major dient.
Minor
	Aktualisierung mit einer abgeschlossenen Verbesserung / Neuerung kleineren Umfanges
Patch
	Bug-Fixes oder kosmetische Änderungen.

## [Unreleased]

## [0.0.0] - YYYY-MM-DD
### Added 
- for new features.
### Changed 
- for changes in existing functionality.
### Deprecated 
- for soon-to-be removed features.
### Removed 
- for now removed features.
### Fixed 
- for any bug fixes.
### Security 
- in case of vulnerabilities.

cmp.h
/*
  cmp.h - This file is the header to the Controllino Message Protocol (CMP) implementation
  CONTROLLINO is a programmable PLC based on arduino technology. See https://controllino.biz/
  CMP is intended to remotely control all digital, analog and Relay inputs and outputs via UDP datagram messages. 
  The current implementation focuses on the Controllino MEGA and must be adopted for different devices.
  CMP may however also be used as general purpose message protocol. 

  Created by C.Sauter, June 1, 2017.
  Released into the public domain.
*/

#ifndef CMP_h
#define CMP_h

#define CMP_VERSION 'A'
#define CMP_OFFSET_VERSION  0
#define CMP_OFFSET_REQRES   1
#define CMP_OFFSET_Property 2
#define CMP_OFFSET_NUMBER   3
#define CMP_OFFSET_DATA     5

#define CMP_ERR_NOERR               "VEE00"
#define CMP_ERR_VERSIONMISMATCH     "VEE01"
#define CMP_ERR_SCOPEDIGITALSNUM    "VEE02"
#define CMP_ERR_SCOPERELAYNUM       "VEE03"
#define CMP_ERR_SCOPEREVOLMEM       "VEE04"
#define CMP_ERR_PROPERTYUNKNOWN     "VEE05"
#define CMP_ERR_SCOPEANALOG         "VEE06"
#define CMP_ERR_SCOPEAINPUTS        "VEE07"
#define CMP_ERR_REQUNKNONW          "VEE08"

//Debug decreases Speed extremly, do not forget to disable it!
//#define CMP_DEBUG
 
#ifdef CMP_DEBUG
 #define DEBUG_PRINT(x)  Serial.print (x)
 #define DEBUG_PRINTLN(x)  Serial.println (x)
#else
 #define DEBUG_PRINT(x)
 #define DEBUG_PRINTLN(x)
#endif

#define UNKNOWN_PIN 0xFF
#define CMP_SIZEOFVOLMEM 50
//extern const char relays[];
//extern const char inputs[];
//extern const char analogs[];
//extern const char digitals[];

void cmp_setup(IPAddress ip, byte * mac, unsigned int localPort);
void cmp_core();
uint8_t cmp_getPinMode(uint8_t pin);

#endif


cmp.cpp
 /*
  cmp.cpp - This file implements the Controllino Message Protocol (CMP)
  CONTROLLINO is a programmable PLC based on arduino technology. See https://controllino.biz/
  CMP is intended to remotely control all digital, analog and Relay inputs and outputs via UDP datagram messages. 
  The current implementation focuses on the Controllino MEGA and must be adopted for different devices.
  CMP may however also be used as general purpose message protocol. 

  Created by C.Sauter, June 1, 2017.
  Released into the public domain.
*/

#include <Controllino.h>
#include <SPI.h>                 // needed for Arduino versions later than 0018
#include <Ethernet.h>
#include <EthernetUdp.h>         // UDP library from: bjoern@cs.stanford.edu 12/30/2008
#include <stdio.h>
#include "cmp.h"

// All relay outputs
const char relays[] = {   CONTROLLINO_R0,
                          CONTROLLINO_R1, 
                          CONTROLLINO_R2,
                          CONTROLLINO_R3,
                          CONTROLLINO_R4,
                          CONTROLLINO_R5,
                          CONTROLLINO_R6,
                          CONTROLLINO_R7,
                          CONTROLLINO_R8,
                          CONTROLLINO_R9,
                          CONTROLLINO_R10,
                          CONTROLLINO_R11,
                          CONTROLLINO_R12,
                          CONTROLLINO_R13,
                          CONTROLLINO_R14,
                          CONTROLLINO_R15     };

// All digital inputs (input only)
const char inputs[] = {   CONTROLLINO_I16, 
                          CONTROLLINO_I17, 
                          CONTROLLINO_I18, 
                          CONTROLLINO_IN0, 
                          CONTROLLINO_IN1     };
// All analog inputs
const char analogs[] = {  CONTROLLINO_A0,
                          CONTROLLINO_A1, 
                          CONTROLLINO_A2,
                          CONTROLLINO_A3,
                          CONTROLLINO_A4,
                          CONTROLLINO_A5,
                          CONTROLLINO_A6,
                          CONTROLLINO_A7,
                          CONTROLLINO_A8,
                          CONTROLLINO_A9,
                          CONTROLLINO_A10,
                          CONTROLLINO_A11,
                          CONTROLLINO_A12,
                          CONTROLLINO_A13,
                          CONTROLLINO_A14,
                          CONTROLLINO_A15     };
                          
// All digital input/outputs which are not input only
const char digitals[] = { CONTROLLINO_D0,
                          CONTROLLINO_D1, 
                          CONTROLLINO_D2,
                          CONTROLLINO_D3,
                          CONTROLLINO_D4,
                          CONTROLLINO_D5,
                          CONTROLLINO_D6,
                          CONTROLLINO_D7,
                          CONTROLLINO_D8,
                          CONTROLLINO_D9,
                          CONTROLLINO_D10,
                          CONTROLLINO_D11,
                          CONTROLLINO_D12,
                          CONTROLLINO_D13,
                          CONTROLLINO_D14,
                          CONTROLLINO_D15,
                          CONTROLLINO_D16,
                          CONTROLLINO_D17,
                          CONTROLLINO_D18,
                          CONTROLLINO_D19     };

// Volatile memory to store random data
volatile int16_t volmem[CMP_SIZEOFVOLMEM];

// An EthernetUDP instance to let us send and receive packets over UDP
EthernetUDP Udp;
// buffers for receiving data
char packetBuffer[UDP_TX_PACKET_MAX_SIZE+1];       
// buffer for sending data
char ReplyBuffer[UDP_TX_PACKET_MAX_SIZE+1] = "";   

// setup 
void cmp_setup(IPAddress ip, byte * mac, unsigned int localPort){
  Ethernet.begin(mac, ip);
  Udp.begin(localPort);

  #ifdef CMP_DEBUG
    Serial.print("Controllinio listens for UDP Datagramms on ");
    Serial.print(ip);
    Serial.print(":");
    Serial.println(localPort);
  #endif
  
  DEBUG_PRINTLN(sizeof(inputs));
  // inputs as inputs
  for (int i = 0; i < sizeof(inputs); ++i){
    pinMode(inputs[i], INPUT);
  }
  // relays as outputs
  for (int i = 0; i < sizeof(relays); ++i){
    pinMode(relays[i], OUTPUT);
  }
}

void cmp_core(){
  int packetSize = Udp.parsePacket();
  // if there's data available, read a packet
  if (packetSize) {
    #ifdef DEBUG
      DEBUG_PRINT("Received packet of size ");
      DEBUG_PRINTLN(packetSize);
      DEBUG_PRINT("From ");
      IPAddress remote = Udp.remoteIP();
      for (int i = 0; i < 4; i++) {
        Serial.print(remote[i], DEC);
        if (i < 3) {
          DEBUG_PRINT(".");
        }
      }
      DEBUG_PRINT(", port ");
      DEBUG_PRINTLN(Udp.remotePort());
    #endif

    // clear packet buffer
    memset(packetBuffer,0,sizeof(packetBuffer));
    memset(ReplyBuffer,0,sizeof(ReplyBuffer));
    // read the packet into packetBufffer
    Udp.read(packetBuffer, UDP_TX_PACKET_MAX_SIZE);
    #ifdef DEBUG
      DEBUG_PRINTLN("Contents:");
      DEBUG_PRINTLN(packetBuffer);
    #endif

    // Divide the datagram into its message parts
    char  V      = packetBuffer[CMP_OFFSET_VERSION];            // Version               'V': CMP Version
    char  R      = packetBuffer[CMP_OFFSET_REQRES];             // Request/Response      'S': Set, 'R': Read
    char  P      = packetBuffer[CMP_OFFSET_Property];           // Property              'A': Analog, 'C': Config, 'D': Digital, 'E': Error, 'I': Input, 'R': Relay, 'V': Volatile Memory
    char  N      = (packetBuffer[CMP_OFFSET_NUMBER]-48)*10      // NN: Number, ASCII 2 Byte to Char conversion.
                    + packetBuffer[CMP_OFFSET_NUMBER+1] - 48;   
    char  D      = packetBuffer[CMP_OFFSET_DATA];               // For Setting: 0:low, 1:high For Config I: Input, P: Pullup, O: Output
    char  PMODE  = (D == 'O') ? OUTPUT : ((D == 'P') ? INPUT_PULLUP : INPUT) ;      
    char  STATE  = (D == 'H') ? HIGH : LOW;
    int   MASK    = atoi(&(packetBuffer[CMP_OFFSET_DATA+1]));
    int   MEMMASK    = atoi(&(packetBuffer[CMP_OFFSET_DATA]));

    //  --------------- Invalid CMP Version -----------------
    if(V != CMP_VERSION){
      strcpy(ReplyBuffer, CMP_ERR_VERSIONMISMATCH);
    }
    //  --------------- Valid CMP Version -------------------
    else{
      strcpy(ReplyBuffer, packetBuffer);

      //  --------------- VS  SET ---------------------------
      if(R == 'S'){
        DEBUG_PRINTLN("R == S");

        //  --------------- VSC  SET Config ------------------
        if(P == 'C'){
          DEBUG_PRINTLN("P == C");

          ReplyBuffer[CMP_OFFSET_DATA] = (PMODE == OUTPUT) ? 'O' : ((PMODE == INPUT_PULLUP) ? 'P' : 'I');
          //ReplyBuffer[CMP_OFFSET_DATA+1] = 0;
        
          if(N == 99){
            for(int i = 0; i < sizeof(digitals); ++i){
              if((int)(1 << i) & MASK) pinMode(digitals[i], PMODE);
            }
          }
          else if (N >= 0 && N < sizeof(digitals)){
            pinMode(digitals[N], PMODE);
          }
          else{
            strcpy(ReplyBuffer, CMP_ERR_SCOPEDIGITALSNUM);
          }
        }
        //  --------------- VSD  SET Digital ------------------
        else if(P == 'D'){
          DEBUG_PRINTLN("P == D");

          ReplyBuffer[CMP_OFFSET_DATA] = (STATE == HIGH) ? 'H' : 'L';
          //ReplyBuffer[CMP_OFFSET_DATA+1] = 0;
           
          if(N == 99){
            for(int i = 0; i < sizeof(digitals); ++i){
              if((int)(1 << i) & MASK){
                pinMode(digitals[N], OUTPUT);
                digitalWrite(digitals[i], STATE);
              }
            }
          }
          else if (N >= 0 && N < sizeof(digitals)){
            pinMode(digitals[N], OUTPUT);
            digitalWrite(digitals[N], STATE);
          }
          else{
            strcpy(ReplyBuffer, CMP_ERR_SCOPEDIGITALSNUM);
          }
        }
        //  --------------- VSE  SET Error ----------------
        else if(P == 'E'){
          DEBUG_PRINTLN("P == E");
          
          // ....
        }
        //  --------------- VSR  SET Relay ------------------
        else if(P == 'R'){
          DEBUG_PRINTLN("P == R");

          ReplyBuffer[CMP_OFFSET_DATA] = (STATE == HIGH) ? 'H' : 'L';
          //ReplyBuffer[CMP_OFFSET_DATA+1] = 0;
          
          if(N == 99){
            for(int i = 0; i < sizeof(relays); ++i)
              if((int)(1 << i) & MASK) digitalWrite(relays[i], STATE);
          }
          else if (N >= 0 && N < sizeof(relays)){
            digitalWrite(relays[N], STATE);
          }
          else{
            strcpy(ReplyBuffer, CMP_ERR_SCOPERELAYNUM);
          }
        }
        //  --------------- VSV  SET Volatile Memory ----------
        else if(P == 'V'){
          DEBUG_PRINTLN("P == V");

          if((N >= 0) && (N < CMP_SIZEOFVOLMEM)){
            volmem[N] = MEMMASK;
            DEBUG_PRINTLN(MASK);
          }
          else {
            strcpy(ReplyBuffer, CMP_ERR_SCOPEREVOLMEM);
          }
        }
        else {
          strcpy(ReplyBuffer, CMP_ERR_PROPERTYUNKNOWN);
        }
      }

      //  --------------- VS  READ --------------------------
      else if(R == 'R'){
        DEBUG_PRINTLN("R == R");

        //  --------------- VRA  READ Analog ----------------
        if(P == 'A'){
          DEBUG_PRINTLN("P == A");
          
          ReplyBuffer[CMP_OFFSET_DATA] = 0;
          char str[15];
          
          if(N == 99){
            for(int i = 0; i < sizeof(analogs); ++i){
               
               if(i > 0) strcat(ReplyBuffer,".");
               
               uint16_t ar;
               for(int rr = 0; rr < 5; ++rr) ar = analogRead(analogs[i]);
               for(int i = 0; i < 15; ++i) str[i] = 0;
               sprintf(str, "%u", ar);
               DEBUG_PRINTLN(str);
               char highbyte = (ar >> 8);
               char lowbyte  = (char)ar;
               strcat(ReplyBuffer,str);
            }
          }
          else if ((N >= 0) && (N < sizeof(analogs))){
            uint16_t ar;
            for(int rr = 0; rr < 5; ++rr) ar = analogRead(analogs[N]);
            sprintf(str, "%u", ar);
            strcat(ReplyBuffer,str);
          }
          else{
            strcpy(ReplyBuffer, CMP_ERR_SCOPEANALOG);
          }
        }
        //  --------------- VRC  READ Config -----------------
        else if(P == 'C'){
          DEBUG_PRINTLN("P == C");
          
          if(N == 99){
            uint32_t bitpattern = 0;
            char str[sizeof(digitals)+1];
            str[sizeof(digitals)] = 0;
            DEBUG_PRINTLN(sizeof(digitals));
            for(int i = 0; i < sizeof(digitals); ++i){
              str[i] = cmp_getPinMode(digitals[i]);
            }
            ReplyBuffer[CMP_OFFSET_DATA] = 0;
            strcat(ReplyBuffer,str);
          }
          else if ((N >= 0) && (N < sizeof(digitals))){
            ReplyBuffer[CMP_OFFSET_DATA] = cmp_getPinMode(digitals[N]);
            ReplyBuffer[CMP_OFFSET_DATA+1] = 0;
          }
          else{
            strcpy(ReplyBuffer, CMP_ERR_SCOPEDIGITALSNUM);
          }
        }
        //  --------------- VRD  READ Digital ----------------
        else if(P == 'D'){
          DEBUG_PRINTLN("P == D");
          
          if(N == 99){
            uint32_t bitpattern = 0;
            DEBUG_PRINTLN(sizeof(digitals));
            for(int i = 0; i < sizeof(digitals); ++i){
              uint32_t dr = digitalRead(digitals[i]);
              bitpattern |= (uint32_t)(dr << i);
            }
            uint16_t highword = (bitpattern >> 16);
            uint16_t lowword  = (uint16_t)bitpattern;
            #ifdef DEBUG
              Serial.print(highword, BIN);
              Serial.println(lowword, BIN);
            #endif
            char str[15];
            ReplyBuffer[CMP_OFFSET_DATA] = 0;
            sprintf(str, "%u", highword);
            strcat(ReplyBuffer,str);
            strcat(ReplyBuffer,".");
            sprintf(str, "%u", lowword);
            strcat(ReplyBuffer,str);
          }
          else if ((N >= 0) && (N < sizeof(digitals))){
            ReplyBuffer[CMP_OFFSET_DATA] = (digitalRead(digitals[N]) == HIGH ? 'H' : 'L');
            ReplyBuffer[CMP_OFFSET_DATA+1] = 0;
          }
          else{
            strcpy(ReplyBuffer, CMP_ERR_SCOPEDIGITALSNUM);
          }
        }
        //  --------------- VRE  READ Error ----------------
        else if(P == 'E'){
          DEBUG_PRINTLN("P == E");
          
          // ....
        }
        //  --------------- VRI  READ Input ----------------
        else if(P == 'I'){
          DEBUG_PRINTLN("P == I");

          if(N == 99){
            char bitpattern = 0;  
            for(int i = 0; i < sizeof(inputs); ++i){
              bitpattern |= (digitalRead(inputs[i]) << i);
            }
            char str[15];
            sprintf(str, "%u", bitpattern);
            ReplyBuffer[CMP_OFFSET_DATA] = 0;
            strcat(ReplyBuffer,str);
          }
          else if (N == 0 || N == 1 || N == 16 || N == 17 || N == 18){
            int index = 0;
            if(N == 16) index = 0;
            else if(N == 17) index = 1;
            else if(N == 18) index = 2;
            else if(N == 0)  index = 3;
            else if(N == 1)  index = 4;
            
            ReplyBuffer[CMP_OFFSET_DATA] = (digitalRead(inputs[index]) == HIGH ? 'H' : 'L');
            ReplyBuffer[CMP_OFFSET_DATA+1] = 0;
          }
          else{
            strcpy(ReplyBuffer, CMP_ERR_SCOPEAINPUTS);
          }
        }
        //  --------------- VRR  READ Relay ----------------
        else if(P == 'R'){
          DEBUG_PRINTLN("P == R");
          
          uint16_t bitpattern = 0;
          
          DEBUG_PRINTLN(sizeof(relays));

          if(N == 99) {
            for(int i = 0; i < sizeof(relays); ++i){
              uint16_t rr = digitalRead(relays[i]);
              bitpattern |= (uint16_t)(rr << i);
            }
            #ifdef DEBUG
              Serial.print(bitpattern, BIN);
            #endif
            char str[15];
            ReplyBuffer[CMP_OFFSET_DATA] = 0;
            sprintf(str, "%u", bitpattern);
            strcat(ReplyBuffer,str);
          }
          else if ((N >= 0) && (N < sizeof(relays))){
            ReplyBuffer[CMP_OFFSET_DATA] = (digitalRead(relays[N]) == HIGH ? 'H' : 'L' );
            ReplyBuffer[CMP_OFFSET_DATA+1] = 0;
          }
          else{
            strcpy(ReplyBuffer, CMP_ERR_SCOPERELAYNUM);
          }
        }
        //  --------------- VRV  READ Volatile Memory --------
        else if(P == 'V'){
          DEBUG_PRINTLN("P == V");
          
          DEBUG_PRINTLN(CMP_SIZEOFVOLMEM);
          
          if((N >= 0) && (N < CMP_SIZEOFVOLMEM)){
            char str[15];
            ReplyBuffer[CMP_OFFSET_DATA] = 0;
            sprintf(str, "%+d", volmem[N]);
            strcat(ReplyBuffer,str);
          }
          else {
            strcpy(ReplyBuffer, CMP_ERR_SCOPEREVOLMEM);
          }
        }
        else {
          strcpy(ReplyBuffer, CMP_ERR_PROPERTYUNKNOWN);
        }
      }
      else {
        strcpy(ReplyBuffer, CMP_ERR_REQUNKNONW);
      }
    }
    
    ReplyBuffer[0] = CMP_VERSION;
    DEBUG_PRINT("ReplyBuffer : ");
    DEBUG_PRINTLN(ReplyBuffer);
    Udp.beginPacket(Udp.remoteIP(), Udp.remotePort());
    Udp.write(ReplyBuffer);
    Udp.endPacket();
  }
}

uint8_t cmp_getPinMode(uint8_t pin)
{
  uint8_t bit = digitalPinToBitMask(pin);
  uint8_t port = digitalPinToPort(pin);

  // I don't see an option for mega to return this, but whatever...
  if (NOT_A_PIN == port) return UNKNOWN_PIN;

  // Is there a bit we can check?
  if (0 == bit) return UNKNOWN_PIN;

  // Is there only a single bit set?
  if (bit & bit - 1) return UNKNOWN_PIN;

  volatile uint8_t *reg, *out;
  reg = portModeRegister(port);
  out = portOutputRegister(port);

  if (*reg & bit)
    //return OUTPUT;
    return 'O';
  else if (*out & bit)
    //return INPUT_PULLUP;
    return 'P';
  else
    //return INPUT;
    return 'I';
}
